### PR TITLE
Add message type to associate CW contract address

### PR DESCRIPTION
**Pull request by codchen**
## Describe your changes and provide context
Add a new `AssociateContractAddress` message type for CW contracts to be explicitly associated with their EVM address counterparts.

## Testing performed to validate your change
unit tests

local Sei
```
tonychen@Tonys-MacBook-Pro sei-chain % ~/go/bin/seid tx evm associate-contract-address sei14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sh9m79m --from=admin --fees=200000usei -y --broadcast-mode=block
code: 0
codespace: ""
data: 0A370A352F73656970726F746F636F6C2E736569636861696E2E65766D2E4D73674173736F6369617465436F6E747261637441646472657373
events:
- attributes:
  - index: true
    key: c3BlbmRlcg==
    value: c2VpMXdncmZ0bHU4bXphOW53cXhjNGxzcHk1dnJncGhjaGxwdDAzemx2
  - index: true
    key: YW1vdW50
    value: MjAwMDAwdXNlaQ==
  type: coin_spent
- attributes:
  - index: true
    key: ZmVl
    value: MjAwMDAwdXNlaQ==
  - index: true
    key: ZmVlX3BheWVy
    value: c2VpMXdncmZ0bHU4bXphOW53cXhjNGxzcHk1dnJncGhjaGxwdDAzemx2
  type: tx
- attributes:
  - index: true
    key: YWNjX3NlcQ==
    value: c2VpMXdncmZ0bHU4bXphOW53cXhjNGxzcHk1dnJncGhjaGxwdDAzemx2LzI=
  type: tx
- attributes:
  - index: true
    key: c2lnbmF0dXJl
    value: c2lyNGducXlkRUFkdncvMlFsdDRUdk55eCt2bEpndXhYakpDM3J5ME9DWUdVUks2Z0hBdE8zNmNRY25lZFdteVJRT2JjcjRRd25nMHFzemhXN0wvbmc9PQ==
  type: tx
- attributes:
  - index: true
    key: ZXZtX2FkZHI=
    value: MHg5NkNkNjhEYkVDNEQ2ZDU3RmVFQkM4MzBmZDA2QThEQWU2ODNBMTc2
  - index: true
    key: c2VpX2FkZHI=
    value: c2VpMXdncmZ0bHU4bXphOW53cXhjNGxzcHk1dnJncGhjaGxwdDAzemx2
  type: signer
- attributes:
  - index: true
    key: YWN0aW9u
    value: L3NlaXByb3RvY29sLnNlaWNoYWluLmV2bS5Nc2dBc3NvY2lhdGVDb250cmFjdEFkZHJlc3M=
  type: message
- attributes:
  - index: true
    key: c2VpX2FkZHI=
    value: c2VpMTRoajJ0YXZxOGZwZXNkd3h4Y3U0NHJ0eTNoaDkwdmh1anJ2Y21zdGw0enIzdHhtZnZ3OXNoOW03OW0=
  - index: true
    key: ZXZtX2FkZHI=
    value: MHg1QThENjQ4REVFNTdiMmZjOTBEOThEQzE3ZmE4ODcxNTliNjk2Mzhi
  type: address_associated
gas_used: "84894"
gas_wanted: "200000"
height: "33"
info: ""
logs:
- events:
  - attributes:
    - key: sei_addr
      value: sei14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sh9m79m
    - key: evm_addr
      value: 0x5A8D648DEE57b2fc90D98DC17fa887159b69638b
    type: address_associated
  - attributes:
    - key: action
      value: /seiprotocol.seichain.evm.MsgAssociateContractAddress
    type: message
  log: ""
  msg_index: 0
raw_log: '[{"events":[{"type":"address_associated","attributes":[{"key":"sei_addr","value":"sei14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sh9m79m"},{"key":"evm_addr","value":"0x5A8D648DEE57b2fc90D98DC17fa887159b69638b"}]},{"type":"message","attributes":[{"key":"action","value":"/seiprotocol.seichain.evm.MsgAssociateContractAddress"}]}]}]'
timestamp: ""
tx: null
txhash: B93ADACED00C29E91F6F1B5BEC73378BA2A7082E49C0FD86B109774EB03E4C29
tonychen@Tonys-MacBook-Pro sei-chain % ~/go/bin/seid q evm evm-addr sei14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sh9m79m
associated: true
evm_address: 0x5A8D648DEE57b2fc90D98DC17fa887159b69638b
tonychen@Tonys-MacBook-Pro sei-chain % ~/go/bin/seid q evm sei-addr 0x5A8D648DEE57b2fc90D98DC17fa887159b69638b
associated: true
sei_address: sei14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sh9m79m
```

### CODE REVIEW

File: x/evm/keeper/msg_server.go [hunk 1]
```diff
@@ -323,3 +324,22 @@ func (server msgServer) RegisterPointer(goCtx context.Context, msg *types.MsgReg
...
+func (server msgServer) AssociateContractAddress(goCtx context.Context, msg *types.MsgAssociateContractAddress) (*types.MsgAssociateContractAddressResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	addr := sdk.MustAccAddressFromBech32(msg.Address) // already validated
+	// check if address is for a contract
+	if server.wasmViewKeeper.GetContractInfo(ctx, addr) == nil {
+		return nil, errors.New("no wasm contract found at the given address")
+	}
+	evmAddr := common.BytesToAddress(addr)
+	existingEvmAddr, ok := server.GetEVMAddress(ctx, addr)
+	if ok {
```
Comment (philipsu522) [inline x/evm/keeper/msg_server.go:337]: do we need to handle the case where ok == false ?

Comment (codchen) [reply x/evm/keeper/msg_server.go:337]: that's the case where we set the association (because no existing association)

**Comment by stevenlanders** [comment]
if someone tries to send an NFT to a contract address and it is not associated, what do they see?

**Comment by codchen** [comment]
> if someone tries to send an NFT to a contract address and it is not associated, what do they see?

the tx would fail with an error

**Pull request merged**